### PR TITLE
Feat(batching): refactor; troubleshoot back pressure

### PR DESCRIPTION
**Pull request by hrmthw**
(Thanks for sending a pull request! Please make sure to read the contribution guidelines, then fill out the blanks below.)

What changes were proposed in this pull request?
------------------------------------------------

Does this close any currently open issues?
------------------------------------------

How was this patch tested?
--------------------------

### CODE REVIEW

File: bentoml/marshal/dispatcher.py [hunk 1]
```diff
@@ -1,176 +1,190 @@
...
-                print("warning: max latency reached")
-                break
-            if max(n - 1, 1) * (wn + dt + a) <= self.outbound_wait * decay:
-                await asyncio.sleep(dt)
-                continue
-            break
-
-    def log_outbound_time(self, info):
-        if info[0] < 5:  # skip all small batch
+        self.o_stat = FixedBucket(self.N_OUTBOUND_SAMPLE)
```
Comment (parano) [inline bentoml/marshal/dispatcher.py:39]: add inline comments here to explain these variables?

File: bentoml/marshal/marshal.py [hunk 2]
```diff
@@ -157,14 +158,17 @@ def set_outbound_port(self, outbound_port):
...
 
     def add_batch_handler(self, api_name, max_latency, max_batch_size):
 
         if api_name not in self.batch_handlers:
             _func = ParadeDispatcher(
-                max_latency, max_batch_size, shared_sema=self.fetch_sema
+                max_latency,
+                max_batch_size,
+                shared_sema=self.fetch_sema(),
+                fallback=aiohttp.web.HTTPTooManyRequests,
```
Comment (parano) [inline bentoml/marshal/marshal.py:171]: Is there a case when `fallback` is not `aiohttp.web.HTTPTooManyRequests`? maybe just set it within ParadeDispatcher for now? something like `self.fallback = aiohttp.web.HTTPTooManyRequests`

File: bentoml/marshal/dispatcher.py [hunk 3]
```diff
@@ -1,176 +1,190 @@
...
+            self._o_b,
+            self._o_w,
+        )
 
 
 class ParadeDispatcher:
-    def __init__(self, max_wait_time, max_size, shared_sema: callable = None):
+    def __init__(
+        self,
+        max_expected_time: int,
```
Comment (parano) [inline bentoml/marshal/dispatcher.py:79]: nit: rename `max_latency_in_ms`?

File: bentoml/marshal/marshal.py [hunk 4]
```diff
@@ -141,6 +141,7 @@ def __init__(
         self.setup_routes_from_pb(self.bento_service_metadata_pb)
         if psutil.POSIX:
             import resource
+
```
Comment (parano) [inline bentoml/marshal/marshal.py:144]: update the docstring in MarshalService class to reflect the changes in this PR?

File: bentoml/utils/alg.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
+import time
+
+
+class FixedBucket:
```
Comment (parano) [inline bentoml/utils/alg.py:4]: consider `deque` with max length for this?
```
from collections import deque
deque(maxlen=100)
```

File: bentoml/marshal/marshal.py (hunk 4 above)
Comment (hrmthw) [reply bentoml/marshal/marshal.py:144]: The usage is almost the same as before.

File: bentoml/marshal/dispatcher.py [hunk 6]
```diff
@@ -1,176 +1,190 @@
...
+        return self.sema < 1
+
+    def release(self):
+        self.sema += 1
 
 
 class Optimizer:
-    N_OUTBOUND_SAMPLE = 500
-    N_OUTBOUND_WAIT_SAMPLE = 20
+    N_OUTBOUND_SAMPLE = 50
```
Comment (parano) [inline bentoml/marshal/dispatcher.py:34]: add inline comments about these variables

File: bentoml/marshal/marshal.py (hunk 4 above)
Comment (parano) [reply bentoml/marshal/marshal.py:144]: I thought we changed the semantic of `mb_max_latency`?

Comment (parano) [reply bentoml/marshal/marshal.py:144]: previously it was closer to `max_wait_time`?

File: bentoml/marshal/marshal.py (hunk 2 above)
Comment (hrmthw) [reply bentoml/marshal/marshal.py:179]: Each request need a new aiohttp.web.HTTPTooManyRequests object or cause exceptions.

File: bentoml/marshal/marshal.py (hunk 4 above)
Comment (hrmthw) [reply bentoml/marshal/marshal.py]: @parano just updated the docstring

File: bentoml/marshal/marshal.py (hunk 2 above)
Comment (hrmthw) [reply bentoml/marshal/marshal.py:179]: The ParadeDispatcher(or CorkDispatcher ?) is just a universal function-wise lib. It's not coupling with http.

Comment (parano) [reply bentoml/marshal/marshal.py:180]: make sense

File: bentoml/utils/alg.py (hunk 5 above)
Comment (parano) [reply bentoml/utils/alg.py:4]: This is not used anymore?

**Pull request merged**